### PR TITLE
30 create user

### PR DESCRIPTION
**Pull request by EugenMakarevich**
Bug 
Description: the user with the existing name and sex is added

Steps to reproduce:
1. Find the existing user
2. Note down the name and sex
3. Create the user with the same name and sex

Expected result: the 400 code response, the user is not added
Actual result: the 201 code response, the user is added

### CODE REVIEW

File: src/main/java/com/coherentsolutions/aqa/java/api/makarevich/service/UserService.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        return generateUser(name, sex);
+    }
+
+    public User createRandomUser(String zipcode) {
+        return generateUser(zipcode);
+    }
+
+    public User createRandomUser() {
+        return generateUser();
+    }
```
Comment (KirKarpov) [inline src/main/java/com/coherentsolutions/aqa/java/api/makarevich/service/UserService.java]: For method that doesn't make API calls to the user service, it is better to make UserFactory class, where you will store all different combinations of user creation. Methods in the factory class usually static
UserService class will have only method that manipulate with user API

File: src/main/java/com/coherentsolutions/aqa/java/api/makarevich/service/UserService.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+    private User generateUser() {
+        return generateUser(null);
+    }
+
+    private UserSex getRandomUserSex() {
+        UserSex[] values = UserSex.values();
+        int index = new Random().nextInt(values.length);
+        return values[index];
+    }
```
Comment (KirKarpov) [inline src/main/java/com/coherentsolutions/aqa/java/api/makarevich/service/UserService.java]: also should be moved to factory class

File: src/main/java/com/coherentsolutions/aqa/java/api/makarevich/service/UserService.java (hunk 1 above)
Comment (KirKarpov) [inline src/main/java/com/coherentsolutions/aqa/java/api/makarevich/service/UserService.java]: I don't think you need such methods if you are not going to have additional logic in them. Currently they just call another private method. If you make method public, most probably you won't need those methods. 
It is up to you, but seems like too many methods may cause confusion in the future

File: src/main/java/com/coherentsolutions/aqa/java/api/makarevich/service/ZipCodeService.java [hunk 3]
```diff
@@ -62,4 +65,8 @@ public boolean isDuplicatedZipcodesPresent(ArrayList<String> list, String zipcod
         }
         return false;
     }
+
+    public String getRandomZipCode() {
+        return faker.address().zipCode();
+    }
```
Comment (KirKarpov) [inline src/main/java/com/coherentsolutions/aqa/java/api/makarevich/service/ZipCodeService.java]: Similar changes will be good to make with ZipCodeService class. I am speaking about moving all methods that related to user creation to factory class. As a result ZipCodeService class will only have methods that interact with API

File: src/test/java/com/coherentsolutions/aqa/java/api/makarevich/CreateUserTest.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        zipCodeService.addZipCode(zipcode);
+        ArrayList<String> zipCodes = zipCodeService.getZipCodes();
+        assertTrue(zipCodes.contains(zipcode), "Zip code is not added");
+
+        User user = userService.createRandomUser(zipcode);
+        userService.createUser(user, SC_CREATED);
+        assertTrue(userService.isUserAdded(user), "User is not added");
+
+        ArrayList<String> zipCodesAfter = zipCodeService.getZipCodes();
+        assertFalse(zipCodesAfter.contains(zipcode), "Zip code is not removed");
```
Comment (KirKarpov) [inline src/test/java/com/coherentsolutions/aqa/java/api/makarevich/CreateUserTest.java]: No need to verify data from each step. For test that check user creation you should have validations is user was created correctly. You expect that other calls are working correctly. If not, there are another tests that should catch it.

**Review by KirKarpov — CHANGES_REQUESTED** [review]

File: src/main/java/com/coherentsolutions/aqa/java/api/makarevich/service/UserService.java (hunk 1 above)
Comment (EugenMakarevich) [reply src/main/java/com/coherentsolutions/aqa/java/api/makarevich/service/UserService.java]: Added UserFactory class and moved it all out of the class

File: src/main/java/com/coherentsolutions/aqa/java/api/makarevich/service/UserService.java (hunk 2 above)
Comment (EugenMakarevich) [reply src/main/java/com/coherentsolutions/aqa/java/api/makarevich/service/UserService.java]: Moved to the UserFactory class

File: src/main/java/com/coherentsolutions/aqa/java/api/makarevich/service/UserService.java (hunk 1 above)
Comment (EugenMakarevich) [reply src/main/java/com/coherentsolutions/aqa/java/api/makarevich/service/UserService.java]: Refactored it, deleted all the methods that call each other and switched to the separate public methods for each case. Moved the code to UserFactory class.

File: src/main/java/com/coherentsolutions/aqa/java/api/makarevich/service/ZipCodeService.java (hunk 3 above)
Comment (EugenMakarevich) [reply src/main/java/com/coherentsolutions/aqa/java/api/makarevich/service/ZipCodeService.java]: Moved to the UserFactory class + isDuplicatedZipcodesPresent() method that does not work with API layer

File: src/test/java/com/coherentsolutions/aqa/java/api/makarevich/CreateUserTest.java (hunk 4 above)
Comment (EugenMakarevich) [reply src/test/java/com/coherentsolutions/aqa/java/api/makarevich/CreateUserTest.java]: Deleted the verification for additional steps, left only verification for the main check for both CreateUserTest and ZipCodeTest

**Review by KirKarpov — APPROVED** [review]

**Pull request merged**